src/error.rs

```diff
@@ -17,12 +18,28 @@ pub enum ConnectionError {
     Checkout(#[from] diesel::r2d2::Error),
 
     #[error("Failed to issue a query: {0}")]
-    Query(#[from] diesel::result::Error),
+    Query(#[from] DieselError),
 }
 
 /// Syntactic sugar around a Result returning an [`PoolError`].
 pub type PoolResult<R> = Result<R, PoolError>;
 
+/// Async varient of
+/// <https://docs.diesel.rs/master/diesel/prelude/trait.OptionalExtension.html>.
+pub trait OptionalExtension<T> {
+    fn optional(self) -> Result<Option<T>, PoolError>;
+}
+
+impl<T> OptionalExtension<T> for PoolResult<T> {
+    fn optional(self) -> Result<Option<T>, PoolError> {
+        match self {
```

Comment:
This looks correct.  Thinking out loud, I wonder if it would be more robust if you used the `OptionalExtension` impl directly, maybe like:

```rust
let self_as_query_result: QueryResult<T> = match self {
    Ok(value) => Ok(value),
    Err(PoolError::Connection(ConnectionError::Query(error_kind)) => Err(error_kind),
    Err(error) => return Err(error)
};

self_as_query_result.optional()
```

but that looks pretty ugly and probably isn't worth it.  (If this were more likely to change, like if there were other DieselError variants that might produce `None` in the future, it might be worth it.)

Reply:
I don't think this can compile - the last case (`Err(error) = > return Err(error)`) tries to convert a `PoolError` directly into a `diesel::result::Error`, which is not possible. `PoolErrors` may contain `diesel::result::Error`, but they also may return errors checking out a connection from the pool itself.